docs/NativeCodeGeneration.md

```diff
@@ -0,0 +1,141 @@
+
+## External - Step 1 - Generation of LLVM IR from high level eBPF program
+
+The first step of the process is identical to the traditional build pipeline. The eBPF program is provided as C or other high level language file and the normal LLVM / Clang tool chain is invoked, to produce a ELF file containing one or more program and zero or more map definitions.
+
+## Step 2 – Verification of the eBPF program
+
+The second step is to use the Prevail verifier to determine if the resulting eBPF program conforms to the required safety requirements. This is identical to the process used to by eBPF for Windows today. Only once the Prevail verifier can prove the program is safe does the process proceed to the next step.
+
+## Step 3 – Generation of the C code from the eBPF byte code
+
+The third step is to generate C code that maps one to one with the eBPF instructions in the ELF file. The eBPF registers are mapped to local variables in the generated C function and each eBPF instruction is used to generate equivalent C operations. In addition, the generated C file also contains bindings for the any helper functions referenced by the eBPF program and any maps the eBPF program uses.
+
+## Step 4 – Compile generate C code into native COFF file
+
+The fourth step is to use the standard C compiler (GCC, MSVC, or Clang) to generate a x64/ARM64 COFF file with Windows calling conventions. The C file will expose a single top-level global variable that allows the skeleton code to find the helpers, maps, and programs contained in the file.
+
+## Step 5 – Linking the COFF file into a PE image
+
+The fifth is to use the platform-specific linker to generate the final PE file. The linker consumed the object file from step 4, a generic DriverEntry skeleton.
+
+## Step 6 – Signing the PE file
+
+The last step is to use the standard PE signing tools to produce a signed PE file from the output of step 5. For the developer workflow the signature would be based on a test signing root, whereas for production use the PE file would be signed using the Microsoft PRS signing keys.
+
+# Stub files consumed by the linker
+
+Step 5 consumes a set of object files to generate the final PE image. The following section gives a of the driver skeleton.
+
+## DriverEntry skeleton
+
+All Windows Drivers need to export a static function labelled &quot;DriverEntry&quot; that is invoked by the OS (Operating System) during loading of the driver. In addition to the standard driver related functionality, the driver entry skeleton would also register three NMR contracts exposed by the other framework components (to export the program to the eBPF runtime, to create any maps required, and to import the helper functions).
+
+## Global entry point for generated code
+
+Every generated C file will contain a single global entry point of type meta_data_table_t:
+
+```
+typedef struct _meta_data_table
```

Comment:
```suggestion
typedef struct _metadata_table
```
metadata is one word.  replace "meta_data" and "meta-data" with "metadata" throughout this doc.